istioctl/cmd/istioctl/metrics.go

```
@@ -0,0 +1,270 @@
+error rate, and request latency at p50, p90, and p99 percentiles. The 
+query results are printed to the console, organized by service name.
+
+All metrics returned are from server-side reports. This means that latencies
+and error rates are from the perspective of the service itself and not of an
+individual client (or aggregate set of clients). Rates and latencies are
+calculated over a time interval of 1 minute.
+`,
+		Example: `
+# Retrieve service metrics for productpage service
+istioctl experimental metrics productpage
+
+# Retrieve service metrics for various services in the different namespaces
+istioctl experimental metrics productpage.foo reviews.bar ratings.baz
+`,
+		Aliases: []string{"m"},
+		Args:    cobra.MinimumNArgs(1),
+		RunE:    run,
+		DisableFlagsInUseLine: true,
+	}
+)
+
+func init() {
+	experimentalCmd.AddCommand(metricsCmd)
+}
+
+type serviceMetrics struct {
+	service                            string
+	totalRPS, errorRPS                 float64
+	p50Latency, p90Latency, p99Latency time.Duration
+}
+
+func run(c *cobra.Command, args []string) error {
+	log.Debugf("metrics command invoked for service(s): %v", args)
+
+	config, _ := defaultRestConfig()
+	client, err := createCoreV1Client()
+	if err != nil {
+		return fmt.Errorf("failed to create k8s client: %v", err)
+	}
+
+	pl, err := prometheusPods(client)
+	if err != nil {
+		return fmt.Errorf("not able to locate prometheus pod: %v", err)
+	}
+
+	if len(pl.Items) < 1 {
+		return errors.New("no prometheus pods found")
+	}
+
+	port, err := availablePort()
+	if err != nil {
+		return fmt.Errorf("not able to find available port: %v", err)
+	}
+	log.Debugf("Using local port: %d", port)
+
+	// only use the first pod in the list
+	promPod := pl.Items[0]
+	fw, readyCh, err := buildPortForwarder(client, config, promPod.Name, port)
+	if err != nil {
+		return fmt.Errorf("could not build port forwarder for prometheus: %v", err)
+	}
```

`defer fw.Close()` here? And ignore the error in case we exited because `fw.ForwardPorts` returned an error.

---

`fw.ForwardPorts()` calls `defer fw.Close()` as the first line in its impl. So, if there is an error with that path, `fw.Close()` will be called.

---

Should `fw.Close()` also be called for other error patches (line 127, line 134)?

---

ok. moved the defer to cover those.